content/sd/0002-apis-de-comunicacao.md

```diff
@@ -12,16 +12,142 @@ type: content
+
+- Os _acknowledgements_ são redundantes, uma vez que os pedidos são seguidos por respostas
+- Estabelecer uma conexão envolve dois pares extra de mensagens para além do par necessário (pedido e resposta)
+- O controlo de fluxo é redundante para a maioria das invocações, visto que usam argumentos e resultados pequenos
+
+:::tip[Nota]
+
+Muitas vezes é difícil decidir o tamanho apropriado para o _buffer_ no qual vamos receber os datagramas (isto aplica-se tanto do lado do cliente como do servidor). O tamanho limitado dos datagramas pode não ser adequado para sistemas [RMI](https://pt.wikipedia.org/wiki/RMI) ou [RPC](/sd/apis-de-comunicacao/#rpc), uma vez que os argumentos e resultados dos _procedures_ podem ter um tamanho arbitrário.
+
 :::
 
+### Exemplo de aplicação UDP
+
+Vamos analisar o comportamento de uma aplicação cliente-servidor que utiliza UDP para a troca de mensagens.
+Ao utilizar UDP, sabemos que as mensagens podem:
+
+- Perder-se
+- Chegar fora de ordem
+- Chegar repetidas
+
+Imaginemos um caso em que o cliente não recebe resposta ao seu pedido. Tem agora 2 opções:
+ou **retorna erro** ou **repete o envio**.
+
+Se chegar uma resposta depois do reenvio, pode ter acontecido 1 de 3 cenários:
+
+- O pedido original não chegou ao servidor (operação executada apenas 1 vez)
+- A resposta ao pedido original perdeu-se (operação executada 2 vezes)
+- O servidor recebeu o pedido original e começou a processá-lo, mas _crashou_, podendo a operação ter sido realizada ou não caso não seja atómica (operação executada 1 ou 2 vezes)
+
+O facto de a operação ser repetida, além de desperdiçar tempo, pode ter efeitos inesperados caso esta não seja **idempotente** (operação que produz o mesmo estado e devolve a mesma resposta independentemente do número de vezes que for executada).
+
+Como é que o servidor pode evitar isto? Tem que ser capaz de verificar se o _id_ do pedido já foi recebido anteriormente:
+
+- Se for a primeira vez: executa a operação normalmente
+- Se for repetido: deve ter guardado o histórico de respostas a pedidos executados e retornar a resposta correspondente (isto implica o armazenamento de _ids_ e suas respostas)
+
+Conseguimos assim perceber que programar sistemas distribuídos através de _sockets_ é bastante complexo e vulnerável a falhas, pelo que iremos utilizar mais um nível de abstração, o RPC.
+
+## RPC
+
+O **RPC** (_Remote Procedure Call_) permite que um cliente execute funções (_procedures_) remotamente num servidor
+mas de forma aparentemente local. O facto da execução ser remota levanta dois problemas principais:
+
+1. A rede entre o cliente e o servidor não é _reliable_, tendo tendência a perder e reordenar mensagens
+2. Os computadores onde ambos os processos correm podem ter arquiteturas muito diferentes
+
+Programação típica com RPC:
+![Programação típica com RPC](./assets/0002-rpc-programming.png#dark=3)
```

Comment:
Do you have these images as SVGs? I'd prefer it

If you're using draw.io, make sure to export it with the diagram included, so that it can be edited in the future, if needed.

Reply:
Would you like all assets in SVG format or just the ones created with Excalidraw?

Reply:
It's preferable to have SVGs over PNGs whenever possible, since they look much better (no detail is lost when scaling).
If you don't have a certain asset as an SVG it's fine, but if you're creating them yourself, aim to export them as SVGs :)